frame/support/src/inherent.rs

```diff
@@ -36,6 +36,11 @@ pub trait ProvideInherent {
 	const INHERENT_IDENTIFIER: self::InherentIdentifier;
 
 	/// Create an inherent out of the given `InherentData`.
+	///
+	/// NOTE: All checks necessary to ensure that the inherent is correct and that can be done in
+	/// the runtime should happen here.
+	/// E.g. if this provides a block producer, check here that the producer is eligible to create
+	/// the block.
```

Comment:
I don't think it is correct, AFAIK `create_extrinsics` is only called by the block producer to get the inherent to include in the block. Nothing prevent the block producer to not call it and include whatever inherent they wants, `create_inherent` only allow to configure in the runtime the generation of inherent to put in a block.

For instance in timestamp pallet, no check is done at all about the validity of the generated extrinsic, it just read the inherent data provided and generate the extrinsic.

Reply:
So then where would that check happen? Would I need to include a check in `on_finalize` that panics in case the block producer is incorrect? Or is an inherent the wrong way of implementing setting the block producer (and it should e.g. be an unsigned extrinsic)?

Reply:
in the example you give of checking if a block producer can actually produce a block, this check should happen in the extrinsic itself. In `#[pallet::call]`.

And if want to require that the inherent extrinsic is provided then you can check in on_finalize (like timestamp)

Reply:
Maybe I should summarize the process of inherent extrinsics. AFAIK this is how it works:

Block producer generate some inherent data. This is coded in the client.
Block producer calls create_inherent with those inherent data in order to get the inherent extrinsics to put into the block.
The block producer produce the block by first including those inherent extrinsics and then other extrinsics.
The block producer give the block to other people.
Other validators will check the block with the api `Core::execute_block`, generate some inherent data, check the inherent of the block with those inherent data using the api `BlockBuilder::check_inherents`.

If you are just a node syncing or maybe in other context I don't know, you will not check the inherents of the blocks.

Reply:
Ok, thanks!
I think I understand better and will reword.